File: 2021/Day10.py
def main():
    with open("input.txt") as f:
        input = f.read()
        print(f"Part 1 Answer: {part1(input)}")
        print(f"Part 2 Answer: {part2(input)}")

def part1(input):
    open = ["(", "[", "{", "<"]
    close = [")", "]", "}", ">"]
    scores = {")": 3, "]": 57, "}": 1197, ">": 25137}
    total = 0
    for line in input.split("\n"):
        prevLine = ""
        while line != prevLine:
            prevLine = line
            for i in range(1, len(line)):
                if line[i-1] in open and line[i] in close and open.index(line[i-1]) == close.index(line[i]):
                    line = line[:i-1] + line[i+1:]
                    break
        for char in line:
            if char not in open:
                total += scores[char]
                break
    return total

def part2(input):
    open = ["(", "[", "{", "<"]
    close = [")", "]", "}", ">"]
    scores = []
    for line in input.split("\n"):
        prevLine = ""
        while line != prevLine:
            prevLine = line
            for i in range(1, len(line)):
                if line[i-1] in open and line[i] in close and open.index(line[i-1]) == close.index(line[i]):
                    line = line[:i-1] + line[i+1:]
                    break
        corrupt = False
        for char in line:
            if char not in open:
                corrupt = True
                break
        if corrupt:
            continue
        scores.append(0)
        for char in line[::-1]:
            scores[-1] *= 5
            scores[-1] += open.index(char) + 1
    scores.sort()
    return scores[len(scores) // 2]

if __name__ == '__main__':
    main()

File: 2024/Day10.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <unordered_set>

int part1(const std::vector<std::string> &input)
{
    int scores = 0;
    for (int i = 0; i < input.size(); i++)
    {
        for (int j = 0; j < input[i].size(); j++)
        {
            if (input[i][j] == '0')
            {
                std::unordered_set<int> visited;
                std::vector<int> frontier;
                frontier.push_back(i * input[i].size() + j);
                while (!frontier.empty())
                {
                    int current = frontier.back();
                    frontier.pop_back();
                    if (visited.find(current) != visited.end())
                    {
                        continue;
                    }
                    visited.insert(current);
                    int x = current / input[i].size();
                    int y = current % input[i].size();
                    int height = input[x][y] - '0';
                    if (height == 9)
                    {
                        scores++;
                        continue;
                    }
                    for (int k = -1; k <= 1; k++)
                    {
                        if (x + k < 0 || x + k >= input.size())
                        {
                            continue;
                        }
                        if (input[x + k][y] - '0' == height + 1)
                        {
                            frontier.push_back((x + k) * input[i].size() + y);
                        }
                    }
                    for (int k = -1; k <= 1; k++)
                    {
                        if (y + k < 0 || y + k >= input[i].size())
                        {
                            continue;
                        }
                        if (input[x][y + k] - '0' == height + 1)
                        {
                            frontier.push_back(x * input[i].size() + y + k);
                        }
                    }
                }
            }
        }
    }
    return scores;
}

int part2(const std::vector<std::string> &input)
{
    int scores = 0;
    for (int i = 0; i < input.size(); i++)
    {
        for (int j = 0; j < input[i].size(); j++)
        {
            if (input[i][j] == '0')
            {
                std::vector<int> frontier;
                frontier.push_back(i * input[i].size() + j);
                while (!frontier.empty())
                {
                    int current = frontier.back();
                    frontier.pop_back();
                    int x = current / input[i].size();
                    int y = current % input[i].size();
                    int height = input[x][y] - '0';
                    if (height == 9)
                    {
                        scores++;
                        continue;
                    }
                    for (int k = -1; k <= 1; k++)
                    {
                        if (x + k < 0 || x + k >= input.size())
                        {
                            continue;
                        }
                        if (input[x + k][y] - '0' == height + 1)
                        {
                            frontier.push_back((x + k) * input[i].size() + y);
                        }
                    }
                    for (int k = -1; k <= 1; k++)
                    {
                        if (y + k < 0 || y + k >= input[i].size())
                        {
                            continue;
                        }
                        if (input[x][y + k] - '0' == height + 1)
                        {
                            frontier.push_back(x * input[i].size() + y + k);
                        }
                    }
                }
            }
        }
    }
    return scores;
}

int main()
{
    std::ifstream file("input.txt");
    std::vector<std::string> input;
    std::string line;
    while (std::getline(file, line))
    {
        input.push_back(line);
    }
    file.close();

    fprintf(stdout, "Part 1 Answer: %d\n", part1(input));
    fprintf(stdout, "Part 2 Answer: %d\n", part2(input));

    return 0;
}